crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,540 @@ impl State {
 
         Ok(derivative)
     }
+
+    /// Checks to see if we should short-circuit the iterative calculation of
+    /// the share proceeds when distributing excess idle liquidity. This
+    /// verifies that the ending LP share price is greater than or equal to the
+    /// starting LP share price and less than or equal to the starting LP share
+    /// price plus the minimum tolerance.
+    fn should_short_circuit_distribute_excess_idle_share_proceeds(
+        &self,
+        active_lp_total_supply: FixedPoint, // just the total number of lp shares
+        starting_present_value: FixedPoint,
+        lp_total_supply: FixedPoint, // total lp shares and withdrawal shares - w.s. ready to withraw
+        present_value: FixedPoint,
+    ) -> Result<bool> {
+        if active_lp_total_supply == fixed!(0) {
+            return Err(eyre!("active_lp_total_supply is zero"));
+        }
+        if lp_total_supply == fixed!(0) {
+            return Err(eyre!("lp_total_supply is zero"));
+        }
+        let new_lp_share_price_down = present_value.div_down(active_lp_total_supply);
+        let new_lp_share_price_up = present_value.div_up(active_lp_total_supply);
+        let old_lp_share_price = starting_present_value.div_up(lp_total_supply);
+        let old_lp_share_price_with_tolerance = (fixed!(1e18)
+            + FixedPoint::from(SHARE_PROCEEDS_SHORT_CIRCUIT_TOLERANCE))
+        .mul_div_down(starting_present_value, lp_total_supply);
+
+        Ok(
+            // Ensure that new LP share price is greater than or equal to the
+            // previous LP share price:
+            //
+            // PV_1 / l_1 >= PV_0 / l_0
+            //
+            // NOTE: Round the LHS down to make the check stricter.
+            new_lp_share_price_down >= old_lp_share_price
+            // Ensure that new LP share price is less than or equal to the
+            // previous LP share price plus the minimum tolerance:
+            //
+            // PV_1 / l_1 <= (PV_0 / l_0) * (1 + tolerance)
+            //
+            // NOTE: Round the LHS up to make the check stricter.
+            && new_lp_share_price_up <= old_lp_share_price_with_tolerance,
+        )
+    }
+
+    /// Calculates the amount of withdrawal shares that can be redeemed and
+    /// the share proceeds the withdrawal pool should receive given the
+    /// pool's current idle liquidity. We use the following algorithm to
+    /// ensure that the withdrawal pool receives the correct amount of
+    /// shares to (1) preserve the LP share price and (2) pay out as much
+    /// of the idle liquidity as possible to the withdrawal pool:
+    ///
+    /// 1. If `y_s * t_s <= y_l * t_l` or
```

Comment:
For function-level docstrings (prefaced with `///`) we use latex notation and `$$` to define equations, as well as "`\cdot`" instead of "`*`". For example,

```
/// 1. if $y_s \cdot t_s <= y_l \cot t_l$ or
```

Reply:
I still need to do this.

Reply:
I wanna follow up with this later so this PR is easier to review